File: otel-agent-extension/src/main/scala/io/scalac/mesmer/otel_agent_extension/OtelAgentExtensionInstrumentation.scala
package io.scalac.mesmer.otel_agent_extension

import io.opentelemetry.javaagent.extension.instrumentation.{ TypeInstrumentation, TypeTransformer }
import net.bytebuddy.asm.Advice
import net.bytebuddy.description.`type`.TypeDescription
import net.bytebuddy.matcher.ElementMatcher
import net.bytebuddy.matcher.ElementMatchers.named

class OtelAgentExtensionInstrumentation extends TypeInstrumentation {
  override def typeMatcher(): ElementMatcher[TypeDescription] =
    named("akka.http.scaladsl.HttpExt")

  override def transform(transformer: TypeTransformer): Unit =
    transformer
      .applyAdviceToMethod(
        named("bindAndHandle"),
        this.getClass.getName + "$OtelAgentExtensionAdvice"
      )
}

class OtelAgentExtensionAdvice {

  @Advice.OnMethodEnter(suppress = classOf[Nothing])
  def onEnter(): Unit =
    println("I was there and I was INJECTED by OTEL AGENT")

}

File: otel-agent-extension/src/main/java/io/scalac/mesmer/otel_agent_extension/OtelAgentExtensionInstrumentation.java
package io.scalac.mesmer.otel_agent_extension;

import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.matcher.ElementMatcher;

import static net.bytebuddy.matcher.ElementMatchers.named;


class OtelAgentExtensionInstrumentation implements TypeInstrumentation {

    @Override
    public ElementMatcher<TypeDescription> typeMatcher() {
        return named("akka.http.scaladsl.HttpExt");
    }

    @Override
    public void transform(TypeTransformer typeTransformer) {
        System.out.println(this.getClass().getName());

        typeTransformer.applyAdviceToMethod(
                named("bindAndHandle"),
                "io.scalac.mesmer.otel_agent_extension.OtelAdvice");

        System.out.println("foo2");
    }
}
